src/java/com/puppetlabs/http/client/impl/StreamingAsyncResponseConsumer.java

```diff
@@ -0,0 +1,88 @@
+package com.puppetlabs.http.client.impl;
+
+import org.apache.http.Header;
+import org.apache.http.HeaderElement;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.entity.DeflateDecompressingEntity;
+import org.apache.http.client.entity.GzipDecompressingEntity;
+import org.apache.http.entity.BasicHttpEntity;
+import org.apache.http.nio.IOControl;
+import org.apache.http.nio.client.methods.AsyncByteConsumer;
+import org.apache.http.protocol.HttpContext;
+
+import java.io.IOException;
+import java.io.PipedInputStream;
+import java.io.PipedOutputStream;
+import java.nio.ByteBuffer;
+
+public class StreamingAsyncResponseConsumer extends AsyncByteConsumer<HttpResponse> {
+
+    private volatile HttpResponse response;
+    private volatile PipedOutputStream pos;
+    private volatile Deliverable<HttpResponse> promise;
+    private volatile Promise<IOException> ioExceptionPromise = new Promise<>();
+
+    public void setFinalResult(IOException ioException) {
+        ioExceptionPromise.deliver(ioException);
+    }
+
+    public StreamingAsyncResponseConsumer(Deliverable<HttpResponse> promise) {
+        this.promise = promise;
+    }
+
+    @Override
+    protected void onResponseReceived(final HttpResponse response) throws IOException {
+        PipedInputStream pis = new ExceptionInsertingPipedInputStream(ioExceptionPromise);
+        pos = new PipedOutputStream();
+        pos.connect(pis);
+        HttpEntity modifiedEntity = new BasicHttpEntity();
+        ((BasicHttpEntity) modifiedEntity).setContent(pis);
+        HttpEntity entity = response.getEntity();
```

Comment:
The code here looks pretty similar to what the `ResponseContentEncoding` class in Apache's httpclient library does - https://github.com/apache/httpclient/blob/4.3.5/httpclient/src/main/java/org/apache/http/client/protocol/ResponseContentEncoding.java#L71.  Would it make sense to just call into that code directly in place of the code here?